RFC: Fast, lispy views

mbauman opened the pull request:
This started as a small experiment to see how much overhead there is when SubArrays simply wrap the passed indices… and rapidly snowballed into a complete solution.  It's absolutely amazing how many tools are available now that allow this sort of simplification.
- Use lispy definitions to re-index the parent indices instead of generated functions
- The last parameter is now simply a boolean that specifies if the SubArray supports fast linear indexing
- Merging indices (for linear indexing) will now create a vector of CartesianIndexes instead of a vector of Ints
- `first_index` and `stride1` are now only computed for FastLinear SubArrays; SubArrays that wrap LinearSlow parents now do no extra work on construction.

This should be relatively minimally breaking; the most observable change is the change in meaning of the last parameter.  I temporarily modified the array tests to test more permutations of SubArrays; here's the performance I saw:

![min_relative](https://cloud.githubusercontent.com/assets/154641/13032116/b66febac-d2b4-11e5-91a1-e1b412b9fe70.png)

`runbenchmarks("array", vs = "JuliaLang/julia:master")`
`

mbauman committed:
Allow constructing CartesianIndex by flattening nested indexes

Also take the dimensionality of CartesianIndex into account when computing index lengths and shapes.

mbauman committed:
Refactor SubArrays

* Use lispy definitions to re-index the parent indices instead of generated functions
* The last parameter is now simply a boolean that specifies if the SubArray supports fast linear indexing
* Merging indices (for linear indexing) will now create a vector of CartesianIndexes instead of a vector of Ints
* first_index and stride1 are now only computed for FastLinear SubArrays; SubArrays that wrap LinearSlow parents now do no extra work on construction.

mbauman commented:
cc @timholy

mbauman commented on test/perf/array/indexing.jl:
```diff
@@ -168,7 +168,18 @@ function makearrays{T}(::Type{T}, sz)
     Bit = trues(sz)
-    (A, AF, AS, ASS, Asub, Bit,)
+    # (A, AF, AS, ASS, Asub, Bit,)
+    (Asub, Asub2, Asub3, Asub4, Asub5)
```
TODO: either restore this back to normal or fully incorporate these tests.

mbauman replied:
I've restored it back to the original for now.  It adds quite a bit of time and complexity to these tests… we may eventually want to create a SubArray perf suite since there are so many different performance-sensitive permutations.

JeffBezanson commented on base/subarray.jl:
```diff
@@ -3,30 +3,78 @@
+function SubArray{P, I, N}(::LinearSlow, parent::P, indexes::I, dims::NTuple{N, Int})
+    SubArray{eltype(P), N, P, I, false}(parent, indexes, dims, 0, 0)
+end
+@generated function SubArray{P, I, N}(::LinearFast, parent::P, indexes::I, dims::NTuple{N, Int})
```
It looks like this doesn't actually need to be a generated function?

timholy replied:
We'd have to make `first_index_expr` and `stride1expr` functions that return the number, rather than the expression to compute the number. But this seems likely to be feasible by lispy

JeffBezanson commented:
Awesome! Have not yet read in detail but looks like a significant simplification and net reduction in generated functions.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/15071) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/master/18e8de7/18e8de7_vs_8ffca32.md). cc @jrevels

timholy commented on base/multidimensional.jl:
```diff
@@ -14,19 +14,34 @@ export CartesianIndex, CartesianRange
+(::Type{CartesianIndex})(index::Integer...) = CartesianIndex(index)
+(::Type{CartesianIndex{N}}){N}(index::Integer...) = CartesianIndex(index)
+# Allow passing tuples smaller than N
+@generated function (::Type{CartesianIndex{N}}){N,M}(index::NTuple{M,Integer})
```
This is busted:
```
julia> CartesianIndex{4}((2,2,2))
ERROR: MethodError: no method matching length(::Type{Tuple{Int64,Int64,Int64}})
 in (::Vararg{Any})() at ./multidimensional.jl:27
 in eval(::Module, ::Any) at ./boot.jl:267
```

timholy replied:
Turns out we can get rid of the `@generated` function here too, see 1ed27d621523aad118f9d8459460551881ac9beb (added to #15030). Then this call would be
```jl
(::Type{CartesianIndex{N}}){N,M}(index::NTuple{M,Integer}) = CartesianIndex{N}(fill_to_length(index, 1, Val{N}))
```
This is also a more composable approach, since `fill_to_length` may be useful in a variety of contexts.

For reference:
```jl
julia> @code_llvm fill_to_length((1,2,3), -1, Val{5})

define void @julia_fill_to_length_23994([5 x i64]* sret, [3 x i64]*, i64, %jl_value_t*) #0 {
top:
  %4 = call %jl_value_t*** @jl_get_ptls_states()
  %5 = getelementptr inbounds [3 x i64], [3 x i64]* %1, i64 0, i64 0
  %6 = load i64, i64* %5, align 8
  %7 = getelementptr inbounds [3 x i64], [3 x i64]* %1, i64 0, i64 1
  %8 = load i64, i64* %7, align 8
  %9 = getelementptr inbounds [3 x i64], [3 x i64]* %1, i64 0, i64 2
  %10 = load i64, i64* %9, align 8
  %11 = insertvalue [5 x i64] undef, i64 %6, 0
  %12 = insertvalue [5 x i64] %11, i64 %8, 1
  %13 = insertvalue [5 x i64] %12, i64 %10, 2
  %14 = insertvalue [5 x i64] %13, i64 %2, 3
  %15 = insertvalue [5 x i64] %14, i64 %2, 4
  store [5 x i64] %15, [5 x i64]* %0, align 8
  ret void
}
```
compare against
```jl
julia> @generated function fill_to_length_gen{M,N}(t::NTuple{M}, val, ::Type{Val{N}})
           M > N && error("input tuple has length $M, asked for $N")
           args = [d <= M ? :(t[$d]) : :(val) for d = 1:N]
           :(tuple($(args...)))
       end

julia> @code_llvm fill_to_length_gen((1,2,3), -1, Val{5})

define void @julia_fill_to_length_gen_23846([5 x i64]* sret, [3 x i64]*, i64, %jl_value_t*) #0 {
top:
  %4 = call %jl_value_t*** @jl_get_ptls_states()
  %5 = getelementptr inbounds [3 x i64], [3 x i64]* %1, i64 0, i64 0
  %6 = load i64, i64* %5, align 8
  %7 = insertvalue [5 x i64] undef, i64 %6, 0
  %8 = getelementptr inbounds [3 x i64], [3 x i64]* %1, i64 0, i64 1
  %9 = load i64, i64* %8, align 8
  %10 = insertvalue [5 x i64] %7, i64 %9, 1
  %11 = getelementptr inbounds [3 x i64], [3 x i64]* %1, i64 0, i64 2
  %12 = load i64, i64* %11, align 8
  %13 = insertvalue [5 x i64] %10, i64 %12, 2
  %14 = insertvalue [5 x i64] %13, i64 %2, 3
  %15 = insertvalue [5 x i64] %14, i64 %2, 4
  store [5 x i64] %15, [5 x i64]* %0, align 8
  ret void
}

timholy replied:
Though it's worse when you exceed 8 args :frowning:

mbauman replied:
Ah, good catch, thank you.  I just made the immediate fix for now, but it really is remarkable how just a few intrinsics allow for more powerful generic code.

timholy commented on base/multidimensional.jl:
```diff
@@ -14,19 +14,34 @@ export CartesianIndex, CartesianRange
     :(CartesianIndex(tuple($(args...))))
 end
-(::Type{CartesianIndex{N}}){M,N}(index::NTuple{M,Integer}) = CartesianIndex{N}(index...)
+# Un-nest passed CartesianIndexes
```
This too could be done without the `@generated` function, although again it will run into trouble for too many args.

timholy commented on base/multidimensional.jl:
```diff
@@ -364,116 +383,57 @@ end
-    N > 0 || throw(ArgumentError("cannot merge empty indexes"))
-    lengthexpr = linindex == Colon ? (:(prod(size(V)[lindim:end]))) : (:(length(linindex)))
-    L = symbol(string("Istride_", N+1))  # length of V's trailing dimensions
+# It returns a CartesianIndex or array of CartesianIndexes.
```
:+1:. It could be more memory-intensive than the original array, but when we get `ReshapedArrays` then we can get rid of this entirely.

timholy replied:
Just in case the memory issue is not apparent:
```jl
A = rand(10,10,10)
B = sub(A, 2:999)
```
`A` consumes 8000 bytes. In `master`, `B` is about the same size (because `sizeof(Int) == sizeof(Float64)`. In this PR, `B` consumes approximately 24000 bytes, so the view is considerably larger than the original.

mbauman replied:
Yeah, I agree that's a problem, but it only manifests itself when taking views of views.

Part of the impetus in refactoring the merge_indexes methods, however, was to make it easier to extend the poly-algorithm in a type-stable manner.  I haven't tried it yet, but I *think* that we can restore the old methods (returning a vector of integer linear indexes) for LinearFast SubArrays.  I think we could even do fancier things when merging, e.g., two colons.

timholy commented on base/subarray.jl:
```diff
@@ -3,30 +3,78 @@
-typealias StridedVector{T,A<:DenseArray,I<:Tuple{Vararg{RangeIndex}}}  Union{DenseArray{T,1}, SubArray{T,1,A,I}}
-typealias StridedMatrix{T,A<:DenseArray,I<:Tuple{Vararg{RangeIndex}}}  Union{DenseArray{T,2}, SubArray{T,2,A,I}}
+# The NoSlice one-element vector type keeps dimensions without losing performance
+immutable NoSlice <: AbstractVector{Int}
```
I'm really glad you agree that this is a better approach. Makes life _so_ much easier!

mbauman replied:
I hope you see your name written in the credits here — I certainly wouldn't have been able to do this from scratch.  :)

timholy commented on base/subarray.jl:
```diff
@@ -3,30 +3,78 @@
 typealias StridedVecOrMat{T} Union{StridedVector{T}, StridedMatrix{T}}
 
+# This computes the linear indexing compatability for a given tuple of indices
+viewindexing() = LinearFast()
```
Ahhh, much better.

timholy commented on base/subarray.jl:
```diff
@@ -3,30 +3,78 @@
-# linear indexing.
-immutable SubArray{T,N,P<:AbstractArray,I<:Tuple{Vararg{ViewIndex}},LD} <: AbstractArray{T,N}
+# L is true if the view itself supports fast linear indexing
+immutable SubArray{T,N,P,I,L} <: AbstractArray{T,N}
```
Do we even still need the `L` parameter, or is `linearindexing` now computable from `I`? Though perhaps it eases dispatch to be able to use a `FastSubArray`?

mbauman replied:
I played with a few permutations here… this one seemed to be the easiest.  The actual computation on `I` is pretty complicated, and it is definitely nice to be able to dispatch on the linear indexability.

Another path I tried (unsuccessfully) was to support strided indexing into any fast parent array ([mb/lispystrided](https://github.com/JuliaLang/julia/compare/master...mbauman:mb/lispystrided)).  Eg., you would then be able to index `(2:5, :, :, :)` with just two indices very rapidly (instead of requiring the full dimensionality of `LinearSlow`).  In general, however, the extra work doesn't seem to be worth it… unless I'm missing something.

timholy commented on base/subarray.jl:
```diff
@@ -347,71 +211,37 @@ first(::Colon) = 1
-    if LD == 0
-        return 0
-    end
+function stride1expr(Itypes, Aexpr, Isym)
```
I assume this could be lisp-ified in a way that inlines at runtime?

mbauman replied:
Ah, yes, I had some un-committed work that didn't make it into this patch that does so.

timholy commented:
Especially if we can get rid of the `@generated` constructors altogether, I'm going to be *really* interested to compare the runtime of `make test-subarray` between master and this PR. (I'm betting you'll cut it by half at least.)

So very, very nice. Thanks for tackling this!

mbauman committed:
Remove the last generated SubArray method

Make the stride1 and first index computations recursive

mbauman committed:
Fix and test CartesianIndex{N} constructors

mbauman committed:
Restore array perf tests

mbauman commented:
> (I'm betting you'll cut it by half at least)

There is a difference, but it's not quite that dramatic… about 40 seconds faster (5%) and 120MB less memory (13%).

timholy commented:
Hmm, I was definitely hoping for more, but it's still very much in the right direction. Plus shorter, more maintainable, and (presumably) statically-compilable (should make @vtjnash happy).

LGTM. I'll leave it up to you to decide whether you want to wait for comments from others, but I am now fine with merging whenever you are.

tkelman commented on test/subarray.jl:
```diff
@@ -89,53 +89,50 @@ function single_stride_dim(A::Array)
-        end
-    end
-end
+# # Compare the linear indexing dimension of a SubArray
```
are these going to be useful again later, or should they just be deleted? it's in the git history and can be brought back from there if needed

mbauman replied:
Ah, I forgot about this chunk.  Yes, it can be deleted — it's intrinsically tied to the meaning of the old `LD` parameter, which was pretty complicated.

mbauman committed:
Remove LD-specific test code

No longer relevant after the meaning of SubArray's final type parameter has changed

tkelman commented:
Merge? Any chance of this fixing the error that's being hit in #14991?

timholy commented:
@tkelman, it should, because it incorporates this: https://github.com/JuliaLang/julia/pull/14529#issuecomment-181183314

mbauman commented:
I'm not quite as certain — your patch was incorporated into #14957, which had already been merged before that AppVeyor run.  It's a non-deterministic failure in a test suite without randomness, right?  Given that, I don't think this will fix it directly… but it might work around it.

Yes, this is good to merge.  Let's give it a shot!
